### PR TITLE
Read timestamp(3) correctly from parquet int64 column

### PR DESCRIPTION
**Pull request by sergey-melnychuk**
This draft PR highlights a problem and introduces a hack that can address this specific use case.

Problems: 
- timestamps written with millisecond precision into int64 (not annotated) parquet column read as 1970's date-time
- parquet reader cannot know in advance (if column is not annotated) that regular int64 actually holds a timestamp
- only when table reads 'timestamp(3)' from 'long' it can check that timestamp needs to be converted to microseconds

Schema like the one below works properly and does not need any workarounds on trino side:

```
message spark_schema {
  optional int64 ts (TIMESTAMP(MILLIS,true));
  optional binary id (STRING);
}
```

Yet if the same data is written using the schema below, the `ts` column is interpreted as microsecond-precision timestamp, thus causing this PR to exist at all:

```
message hive_schema {
  optional binary id (STRING);
  optional int64 ts;
}
```

I know this is a workaround (a dirty hack) and I don't expect it to be merged, but I would highly appreciate any feedback on how I can get the same effect (int64 column being read correctly as millisecond-precision timestamp) without workarounds.

The `TestTimestamp` has been split into two test-cases to highlight the problem.

The separate commit with comments in relevant code paths can be easily removed if necessary.

Relevant PR:
https://github.com/trinodb/trino/pull/9330

### CODE REVIEW

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java [hunk 1]
```diff
@@ -28,6 +29,17 @@ public LongColumnReader(RichColumnDescriptor descriptor)
...
     {
-        type.writeLong(blockBuilder, valuesReader.readLong());
+        long longValue = valuesReader.readLong();
+
+        if (type instanceof TimestampType) {
+            int precision = ((TimestampType) type).getPrecision();
+            if (precision == 3) {
+                // When the 'timestamp(3)' (millisecond precision) is read from the column,
+                // just make sure it is converted to microsecond precision, which TimestampType expects
+                longValue *= 1000;
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java:39]: What does the Hive do in such a case?

Add product tests covering this, and proving we are consistent.

**Comment by findepi** [comment]
Relates to https://github.com/trinodb/trino/pull/9330, cc @sbernauer

**Comment by findepi** [comment]
See @martin's comment here: https://github.com/trinodb/trino/issues/5564#issuecomment-710123644

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java [hunk 2]
```diff
@@ -112,12 +112,14 @@ public static PrimitiveColumnReader createReader(RichColumnDescriptor descriptor
...
                 }
                 if (descriptor.getPrimitiveType().getOriginalType() == OriginalType.TIMESTAMP_MILLIS) {
+                    // The 'int64 test (TIMESTAMP_MILLIS)' column is handled here because of the type annotation.
                     return new Int64TimestampMillisColumnReader(descriptor);
                 }
                 if (descriptor.getPrimitiveType().getLogicalTypeAnnotation() instanceof TimestampLogicalTypeAnnotation &&
                         ((TimestampLogicalTypeAnnotation) descriptor.getPrimitiveType().getLogicalTypeAnnotation()).getUnit() == LogicalTypeAnnotation.TimeUnit.NANOS) {
                     return new Int64TimestampNanosColumnReader(descriptor);
                 }
+                // Yet 'int64 test' column will land into LongColumnReader, as parquet reader cannot know the table's column type in advance.
```
[the remark this answers is not in the document]
Comment (sergey-melnychuk) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java:122]: Sure, why not? Are there more such implicitly-convertible types between Trino and Parquet?

**Comment by sergey-melnychuk** [comment]
Just a thought: if we put this (implicit conversion from int64 to timestamp(3)) behind a feature flag (say, session property 'enable-millis-timestamp-from-parquet-int64'), and based on the feature flag decorate LongColumnRerader with something like TimestampAwareLongColumnReader, I believe it is possible to keep the balance between having and not having a (clean and manageable) workaround in the code.
It might unlock frictionless transition for historic data, which might be valuable for users with *lots* of data.

**Comment by sergey-melnychuk** [comment]
Plot twist: One line actually stops `int64` values to be misinterpreted and microsecond-precision timestamps:
https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/type/SqlTimestamp.java#L53

If for example `int64` contains timestamps in milliseconds rounded to seconds (ending with 000), the resulting timestamp(3) is likely to land in 1970's. So it does seem like a bug to me, thus detecting the case when `timestamp(3)` is being read from `long` is still necessary, limiting us to only two options (at least the way I see it): 1) throw exception; or 2) "fix" the value. IMHO the "clean" approach would be to throw an exception and fail the query, while "practical" approach would be to trivially produce necessary **valid** value.

int64-timestamp-meme

**Comment by sergey-melnychuk** [comment]
Suggestion for stricter type/value assertions in Tempto: https://github.com/trinodb/tempto/pull/93

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/Int64TimestampMillisColumnReader.java [hunk 3]
```diff
@@ -44,6 +44,7 @@ protected void readValue(BlockBuilder blockBuilder, Type type)
             type.writeLong(blockBuilder, packDateTimeWithZone(epochMillis, UTC_KEY));
         }
         else if (type instanceof TimestampType) {
+            // The existing int64-millis-timestamp support is just converting timestamp to microsecond precision.
```
Comment (martint) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/Int64TimestampMillisColumnReader.java:47]: If this comment is meant to explain what the next line does, it seems redundant.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java (hunk 1 above)
Comment (martint) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java:39]: Then, why are we thinking of adding this non-standard (per Hive behavior) conversion?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/Int64TimestampMillisColumnReader.java (hunk 3 above)
Comment (sergey-melnychuk) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/Int64TimestampMillisColumnReader.java:47]: it is, just put some comments in relevant places to highlight key timestamp handling steps; 
such comments are in a separate commit that is not intended to reach beyond draft pr.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java (hunk 1 above)
Comment (sergey-melnychuk) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java:39]: Just presenting a way how it *can be* handled. It does seem like a "quick win" that many users might find valuable, or at least a way to address possible nasty bug (see https://github.com/trinodb/trino/pull/10661#issuecomment-1016457070).

**Comment by sergey-melnychuk** [comment]
Test case is ready (https://github.com/trinodb/trino/pull/10661/commits/e55cac8110ccabbfcafb38c508041469a372a2fe) - it shows the behaviour described in (https://github.com/trinodb/trino/pull/10661#issuecomment-1016457070) - misinterpreting milliseconds rounded to seconds as microseconds without the `value *= 1000;` workaround.

The `testSelectPlainInt64MillisTimestamp` passes with or without the workaround, annotation does the job.

The original problem is reproduced in `testSelectAnnotatedInt64MillisTimestamp`:

```
tests               | io.trino.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20220120_105328_00004_np4n9): Could not serialize column 'ts' of type 'timestamp(3)' at position 1:2
<snip>
tests               | Caused by: java.sql.SQLException: Query failed (#20220120_105328_00004_np4n9): Could not serialize column 'ts' of type 'timestamp(3)' at position 1:2
<snip>
tests               | Caused by: java.lang.IllegalArgumentException: Expected 0s for digits beyond precision 3: epochMicros = 1632127786531
```

The bug with milliseconds rounded to seconds is reproduced in `testSelectRoundedInt64MillisTimestamp`:

```
java.lang.AssertionError: Could not find rows:
tests               | [5ab115df-c3cf-470e-966b-66c96f4d2004, 2021-09-20 08:49:46.531]
tests               | [dec75266-3293-471c-9023-74693eb1247e, 2021-09-20 08:49:46.531]
tests               | [a4d7956f-47dc-4674-b1af-f335f6c9ddfc, 2021-09-20 08:49:46.531]
tests               |
tests               | actual rows:
tests               | [5ab115df-c3cf-470e-966b-66c96f4d2004, 1970-01-19 21:22:07.786]
tests               | [dec75266-3293-471c-9023-74693eb1247e, 1970-01-19 21:22:07.786]
tests               | [a4d7956f-47dc-4674-b1af-f335f6c9ddfc, 1970-01-19 21:22:07.786]
```

**Comment by sergey-melnychuk** [comment]
@findepi @martint @electrum should I proceed or drop this? The way I see it this one change (that mostly fits current design) adds a valuable feature and fixes a nasty bug. Please let me know if I am missing some side-effects on unwanted consequences of such change, because at this point I can't see any.

**Comment by k-haley1** [comment]
@findepi @martint @electrum please see Sergey's comments on the best path forward. We need to give the customer an update on this asap.

**Comment by findepi** [comment]
> See @martin's comment here: [#5564 (comment)](https://github.com/trinodb/trino/issues/5564#issuecomment-710123644)

@sergey-melnychuk @martint did you have an explicit conversation about that?

**Pull request closed without merging**